### PR TITLE
Implementing upsert and adding tests

### PR DESCRIPTION
**Pull request by hannelita**
Now we have Mongo docs being inserted as nodes with their respective properties in Neo4j
screen shot 2015-08-11 at 12 25 15 am

All the nodes come labelled as Document:Mongo_collection, as @jexp suggested

### CODE REVIEW

File: mongo_connector/doc_managers/neo4j_doc_manager.py [hunk 1]
```diff
@@ -53,7 +52,26 @@ def upsert(self, doc, namespace, timestamp):
...
   
+  def store_nodes_and_relationships(self, doc, doc_type, doc_id):
+    tx = self.graph.cypher.begin()
+    parameters = {'id':doc_id}
+    for key in doc.keys():
+      LOG.error(type(doc[key]))
+      if (type(doc[key]) is str):
+        value = doc[key]
+      else:
+        value = ""
```
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: just skip empty values in the dict

File: mongo_connector/doc_managers/neo4j_doc_manager.py [hunk 2]
```diff
@@ -53,7 +52,26 @@ def upsert(self, doc, namespace, timestamp):
...
+      parameters.update({ key: value })
+    parameters = ', '.join("{!s}:{!r}".format(k,v) for (k,v) in parameters.items())
+    query = "CREATE (c:Document:{doc_type} {{ {parameters} }}) RETURN c".format(doc_type=doc_type, parameters=parameters)
```
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: don't do this, always use parameters within the statement no string formatting (except for the label)

File: mongo_connector/doc_managers/neo4j_doc_manager.py [hunk 3]
```diff
@@ -53,7 +52,26 @@ def upsert(self, doc, namespace, timestamp):
...
+    tx.append(query)
```
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: tx.append(query,params)

File: mongo_connector/doc_managers/neo4j_doc_manager.py [hunk 4]
```diff
@@ -33,13 +31,14 @@ class DocManager(DocManagerBase):
   def __init__(self, url, auto_commit_interval=DEFAULT_COMMIT_INTERVAL,
                  unique_key='_id', chunk_size=DEFAULT_MAX_BULK, **kwargs):
     self.remote_graph = Graph(url, **kwargs.get('clientOptions', {}))
-    graph = Graph()
+    self.graph = Graph()
     self.auto_commit_interval = auto_commit_interval
     self.unique_key = unique_key
     self.chunk_size = chunk_size
     if self.auto_commit_interval not in [None, 0]:
```
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: no auto-commit in a separate thread, but just after X updates per transaction (X = 10k-50k nodes / rels)

File: mongo_connector/doc_managers/neo4j_doc_manager.py [hunk 5]
```diff
@@ -33,13 +31,14 @@ class DocManager(DocManagerBase):
   def __init__(self, url, auto_commit_interval=DEFAULT_COMMIT_INTERVAL,
                  unique_key='_id', chunk_size=DEFAULT_MAX_BULK, **kwargs):
     self.remote_graph = Graph(url, **kwargs.get('clientOptions', {}))
-    graph = Graph()
+    self.graph = Graph()
```
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: On INIT run `CREATE CONSTRAINT ON (d:Document) ASSERT d.id IS UNIQUE;`

File: mongo_connector/doc_managers/neo4j_doc_manager.py [hunk 6]
```diff
@@ -53,7 +52,26 @@ def upsert(self, doc, namespace, timestamp):
...
+      LOG.error(type(doc[key]))
+      if (type(doc[key]) is str):
+        value = doc[key]
+      else:
+        value = ""
+      parameters.update({ key: value })
+    parameters = ', '.join("{!s}:{!r}".format(k,v) for (k,v) in parameters.items())
+    query = "CREATE (c:Document:{doc_type} {{ {parameters} }}) RETURN c".format(doc_type=doc_type, parameters=parameters)
+    tx.append(query)
+    LOG.error(query)
```
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: logging should be info or debug level in the final one.

File: tests/test_neo4j_doc_manager.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+import sys
+import time
+
+sys.path[0:0] = [""]
+
+from mongo_connector.command_helper import CommandHelper
+from mongo_connector.doc_managers.elastic_doc_manager import DocManager
+
+
+class TestNeo4jDocManager(Neo4jTestCase):
```
Comment (jexp) [inline tests/test_neo4j_doc_manager.py]: is this similar to the other test-cases for other database?

File: tests/test_neo4j_doc_manager.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  def test_update(self):
+    """Test the update method."""
+    doc_id = 1
+    doc = {"_id": doc_id, "speaker": 1, "sessions": 2}
+    self.assertEqual(doc, {"_id": '1', "sessions": 3})
```
Comment (jexp) [inline tests/test_neo4j_doc_manager.py]: shouldn't this use doc_id instead of '1' ?

File: mongo_connector/doc_managers/neo4j_doc_manager.py (hunk 5 above)
Comment (hannelita) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: Questions deriving from here:
- SHould we use Graph or remote graph?
- We will need to make the authentication step. So I'll add these parameters into the constructor also. ( http://py2neo.org/2.0/essentials.html#authentication )

File: mongo_connector/doc_managers/neo4j_doc_manager.py (hunk 4 above)
Comment (hannelita) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: Actually this was the default config. But makes sense changing the default. I'll check how I can make this; since it's managed by time, I think. Not by transanction number

File: mongo_connector/doc_managers/neo4j_doc_manager.py (hunk 1 above)
Comment (hannelita) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: As I mentioned here, https://github.com/hannelita/neo4j_doc_manager/commit/6214d585375b0b6d63c4e030d1296c6a33a1ee48#commitcomment-12627294 , This will be changed for relationship handling

File: mongo_connector/doc_managers/neo4j_doc_manager.py (hunk 2 above)
Comment (hannelita) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: Mentioned here, https://github.com/hannelita/neo4j_doc_manager/commit/6214d585375b0b6d63c4e030d1296c6a33a1ee48#commitcomment-12627313 , theres a problem with space trimming. {'key is what happens. I will find a way to suppress this. So I'll be able to do something like 
self.graph.cypher.execute("CREATE (c:Document:{document_type} { {prop_name}:{N}}) RETURN c", {"N": {speaker_name} }). Note the { {. { { gets trimmed. But there should be a way to space it.

**Comment by jexp** [comment]
Ready to merge this pull request @hannelita ?

**Comment by hannelita** [comment]
@jexp If it seems ok to you, you can merge. I still pushing changes to here because all the other methods depend on this PR

**Comment by hannelita** [comment]
@jexp https://github.com/neo4j-contrib/neo4j_doc_manager/pull/2/files#r36712319 What do you think of these points? I can also add them into this PR or make another one

File: tests/test_neo4j_doc_manager.py (hunk 8 above)
Comment (jexp) [inline tests/test_neo4j_doc_manager.py]: ping?

File: tests/test_neo4j_doc_manager.py (hunk 7 above)
Comment (jexp) [inline tests/test_neo4j_doc_manager.py]: ??

File: mongo_connector/doc_managers/neo4j_doc_manager.py [hunk 9]
```diff
@@ -33,13 +31,20 @@ class DocManager(DocManagerBase):
...
     self.chunk_size = chunk_size
     if self.auto_commit_interval not in [None, 0]:
       self.run_auto_commit()
-  
+    self._formatter = DefaultDocumentFormatter()
+    self.query_nodes = {}
+    self.doc_types = []
+
+  def apply_id_constraint(self, doc_type):
+    constraint = "CREATE CONSTRAINT ON (d:{doc_type}) ASSERT d.id IS UNIQUE".format(doc_type=doc_type)
```
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: As mongodb-id's are unique across all collections we could have the constraint on :Document(id)
It also makes sense to have it on a collection level though as most interaction will go through the collection-node.
This would be up for discussion.

File: mongo_connector/doc_managers/neo4j_doc_manager.py [hunk 10]
```diff
@@ -53,13 +58,53 @@ def upsert(self, doc, namespace, timestamp):
...
+    main_type = self.doc_types.pop(0)
+    for node_type in self.doc_types:
+      tx.append(self.build_relationships(doc_id, main_type, node_type))
+    tx.commit()
+
+  def build_nodes(self, doc_type, hash, id):
+    self.doc_types.append(doc_type)
+    parameters = {'id':id}
+    for key in hash.keys():
+      if (type(hash[key]) is str) or (type(hash[key]) is list):
```
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: booleans and numbers?
list == list of primitives (string,boolean, number) or also list of doc-ids / expanded json-objects ?

File: mongo_connector/doc_managers/neo4j_doc_manager.py [hunk 11]
```diff
@@ -53,13 +58,53 @@ def upsert(self, doc, namespace, timestamp):
...
+        value = hash[key]
+        parameters.update({ key: value })
+      else:
+        self.build_nodes(key, hash[key], id)
+    query = "CREATE (c:Document:{doc_type} {{parameters}})".format(doc_type=doc_type)
+    self.query_nodes.update({query: parameters})
+
+  def build_relationships(self, doc_id, main_type, node_type):
+    relationship_type = main_type + "_" + node_type
+    statement = "MATCH (a:{main_type}), (b:{node_type}) WHERE a.id = '{doc_id}' AND b.id = '{doc_id}' CREATE (a)-[r:{relationship_type}]->(b)".format(main_type=main_type, node_type=node_type, doc_id=doc_id, relationship_type=relationship_type)
```
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: We have to make sure that this match is aligned with the constraint we create, i.e. if we have a constraint on :Document(id) we would have to macht here against :Document(id) otherwise, this match against `:{type}.id` is correct.

Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: Will those relationships only created once? Or is this method also called for updates? I.e. should we make sure to use MERGE or CREATE UNIQUE on rels?

Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: Is it guaranteed that both ends (both nodes) already exist in the db when we create the rels? Otherwise this operation will silently not do anything as it won't find the nodes.
If we can't guarantee that we will have to use MERGE instead of MATCH but that's of course a more expensive operation.

File: mongo_connector/doc_managers/neo4j_doc_manager.py [hunk 12]
```diff
@@ -53,13 +58,53 @@ def upsert(self, doc, namespace, timestamp):
...
 
   def update(self, document_id, update_spec, namespace, timestamp):
-    LOG.error("Update")
+    self.commit()
+    doc_id = u(document_id)
+    update_value_list = update_spec['$set']
+    index, doc_type = self._index_and_mapping(namespace)
+    tx = self.graph.cypher.begin()
+    for update_value in update_value_list.keys():
+      statement = "MATCH (d:Document:{doc_type}) WHERE d.id = '{doc_id}' SET d.{update_value} = '{value}'".format(doc_type=doc_type, doc_id=doc_id, update_value=update_value, value=update_value_list[update_value])
```
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: this updates a singular value at a time?
Did you also consider `SET d += {map-of-updates}`
Also you're not using cypher prameters here, or? For both `doc_id` as well as `value`?

**Comment by jexp** [comment]
Had some further comments before merging, please have a look.
Thanks a lot.

File: mongo_connector/doc_managers/neo4j_doc_manager.py (hunk 9 above)
Comment (hannelita) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: So, ids are unique acress collections. Which lead us to have this, for example
```
{"_id":{"$oid":"55caddc301f4f2aa0b7ea772"},"room":"Auditorium","session":{"abstract":"Spring emerged as a core open source project in early 2003 and evolved to a broad portfolio of open source projects up until 2015. This keynote reflects upon the journey so far, with a focus on the open source ecosystem that Spring lives within, including stories and anecdotes from the old days as well as from recent times. Not getting stuck in history, we’ll also look at the continuity of Spring’s philosophy and its immediate applicability to Java development challenges in 2015 and beyond.","title":"12 Years of Spring: An Open Source Journey"},"speaker":{"bio":"Juergen Hoeller is co-founder of the Spring Framework open source project and has been serving as the project lead and release manager for the core framework since 2003. Juergen is an experienced software architect and consultant with outstanding expertise in code organization, transaction management and enterprise messaging.","name":"Juergen Hoeller","picture":"http://www.springio.net/wp-content/uploads/2014/11/juergen_hoeller-220x220.jpeg","twitter":"https://twitter.com/springjuergen"},"timeslot":"Wed 29th, 09:30-10:30","topics":["keynote","spring"]}
```
There's a id referencing the collection
But three nodes should be created here, a:Document:[collection_name] //in this case, for example, talk, b:document:session, c:Document:speaker. All under the same Mongo ID. So the nodes of the collection will carry the same ID. Wouldnt it cause conflicts?

File: mongo_connector/doc_managers/neo4j_doc_manager.py (hunk 10 above)
Comment (hannelita) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: As far as I understood, self._formatter.format_document(doc) will strignify the content, so id : 1234 becomes id: '1234'

File: mongo_connector/doc_managers/neo4j_doc_manager.py (hunk 11 above)
Comment (hannelita) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: - upsert is not called by update, unless it is called by mongo itself
- this method is not called for updates in its raw form, only if mongo calls an upsert. For example, $set { abc: {def:ghi, jkl:mno} } and this subhash does not exist, it will trigger mongo upsert.

File: mongo_connector/doc_managers/neo4j_doc_manager.py (hunk 5 above)
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: ya auth. IDK the difference between graph or remote graph we should ask @nigel

File: mongo_connector/doc_managers/neo4j_doc_manager.py (hunk 9 above)
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: From our discussion I understood, that when the nested documents are real documents then their id's will be provided too.

Otherwise if they are just embeded data pieces, we have to configure which of their fields has to be used as unique keys (e.g. in this example speaker.twitter) . Or we have to artificially create keys by parent-id-field-name-[field-array-index]

File: mongo_connector/doc_managers/neo4j_doc_manager.py (hunk 10 above)
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: We should keep the original datatype, not change everything to string.

Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: the question was also more about the if condition, where strings and lists are used as parameters, and all other datatypes will cause the call to `build_nodes`

File: mongo_connector/doc_managers/neo4j_doc_manager.py (hunk 11 above)
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: sorry, I couldn't follow. my question was about the graph structure when that method is called. if both nodes that we try to connect are guaranteed there or not :)

File: mongo_connector/doc_managers/neo4j_doc_manager.py (hunk 9 above)
Comment (hannelita) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: Ok, this case will also happen. But take a look at the above example. By inserting this kind of document, we would have a bunch of nested data. Not always we will have a proper id for then. Keeping it into the same node wouldn't be bad?

File: mongo_connector/doc_managers/neo4j_doc_manager.py (hunk 10 above)
Comment (hannelita) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: Yes, I got the point. This is being done to split dictionaries into other nodes. But refer this comment - https://github.com/neo4j-contrib/neo4j_doc_manager/pull/2/files#r36886511 - Should we make it? I think so, even if it's something without a proper id.
- if you think we should not split this into a separate node, I can change the logic and consider a nested id only
- if we consider splitting, I can a. remove the formatter provided by the connector and b. inverse the logic and consider dictionaries

File: mongo_connector/doc_managers/neo4j_doc_manager.py [hunk 13]
```diff
@@ -33,13 +31,20 @@ class DocManager(DocManagerBase):
...
-    graph = Graph()
+    self.graph = Graph()
     self.auto_commit_interval = auto_commit_interval
     self.unique_key = unique_key
     self.chunk_size = chunk_size
     if self.auto_commit_interval not in [None, 0]:
       self.run_auto_commit()
-  
+    self._formatter = DefaultDocumentFormatter()
+    self.query_nodes = {}
```
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: Is there a new instance of doc-manager created for each update call? And what about concurrency on these calls?
Otherwise the variables should probably be method local and passed on?

File: mongo_connector/doc_managers/neo4j_doc_manager.py [hunk 14]
```diff
@@ -53,16 +58,73 @@ def upsert(self, doc, namespace, timestamp):
...
+    for node_type in self.doc_types:
+      tx.append(self.build_relationships(main_type, node_type), {"doc_id": doc_id})
+    tx.commit()
+    self.query_nodes = {}
+    self.doc_types = []
+
+  def build_nodes(self, doc_type, hash, id):
+    self.doc_types.append(doc_type)
+    parameters = {'_id':id}
+    for key in hash.keys():
```
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: this should be made more explicit, i.e. the checks should be method names that contain the conditionals, like "is_simple_property", "is_relationship_property", "is_many_relationships_property"

File: mongo_connector/doc_managers/neo4j_doc_manager.py [hunk 15]
```diff
@@ -53,16 +58,73 @@ def upsert(self, doc, namespace, timestamp):
...
+        for json in hash[key]:
+          json_key = key + str(hash[key].index(json))
+          self.build_nodes(json_key, json, id)
+      else:
+        value = hash[key]
+        parameters.update({ key: value })
+    query = "CREATE (c:Document:{doc_type} {{parameters}})".format(doc_type=doc_type)
+    self.query_nodes.update({query: parameters})
+
+  def build_relationships(self, main_type, node_type):
```
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: build_relationship_query

File: mongo_connector/doc_managers/neo4j_doc_manager.py [hunk 16]
```diff
@@ -53,16 +58,73 @@ def upsert(self, doc, namespace, timestamp):
...
+    tx.commit()
 
   def remove(self, document_id, namespace, timestamp):
-    LOG.error("remove")
+    """Removes a document from Neo4j."""
+    doc_id = u(document_id)
+    index, doc_type = self._index_and_mapping(namespace)
+    params_dict = {"doc_id": doc_id}
+    tx = self.graph.cypher.begin()
+    statement = "MATCH (d:Document) WHERE d._id={doc_id} OPTIONAL MATCH (d)-[r]-() DELETE d, r"
```
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: if we don't have an index on `:Document(_id)` this will not be a fast query

Comment (hannelita) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: Agree, but deletes may be slow in Mongo. So here comes the question: Does it makes sense to have an index here because of the removal? 
I can see the point. Should it be added into this PR?

File: mongo_connector/doc_managers/neo4j_doc_manager.py (hunk 15 above)
Comment (hannelita) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: I thought that refactor points should be added into a separate PR (i have started one for tests, general refactor and file split)

File: mongo_connector/doc_managers/neo4j_doc_manager.py (hunk 13 above)
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: ??

File: mongo_connector/doc_managers/neo4j_doc_manager.py (hunk 15 above)
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: ok :)

File: mongo_connector/doc_managers/neo4j_doc_manager.py (hunk 16 above)
Comment (jexp) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: Still in a large graph this operation would then be O(n) which we definitely can't afford, also those deletes will block other operations.

You don't have to add it to this PR. We just must not forget about it.

File: mongo_connector/doc_managers/neo4j_doc_manager.py (hunk 13 above)
Comment (hannelita) [inline mongo_connector/doc_managers/neo4j_doc_manager.py]: I'm refactoring this to add query_nodes for each upsert call
